source/opt/instruction.h

```diff
@@ -246,6 +254,39 @@ inline void Instruction::ForEachInst(
   f(this);
 }
 
+inline void Instruction::ForEachId(const std::function<void(uint32_t*)>& f) {
+  for (auto& opnd : operands_) {
+    switch (opnd.type) {
+      case SPV_OPERAND_TYPE_RESULT_ID:
+        f(&opnd.words[0]);
+        result_id_ = opnd.words[0];
```

Comment:
It needs to set `result_id_` and `type_id_`, as otherwise if you later call `Instruction::result_id()`, you will still get the old value, which is probably not what you would expect as you changed it in the call to `Instruction::ForEachId()`.

However, I do agree that the code could be simplified quite a bit, and look like:
```
for (auto operand : operands_) {
  if (spvIsIdType(operand.type)) f(operand.words[0]);
}
if (result_id_ != 0u) result_id_ = GetSingleWordInOperand(0);
if (type_id_ != 0u) type_id_ = GetSingleWordInOperand(1);
```
And that code should probably check that `f()` did not cause `result_id_` and `type_id_` to equal 0, as that would confuse the `Instruction` class as it relies on them being null to know whether they are set or not.